code/datums/elements/area_based_godmode.dm

```diff
@@ -0,0 +1,83 @@
+/**
+ * Area-based godmode.
+ * (area, allow_area_subtypes, gain_message, lose_message)
+ */
+/datum/element/area_based_godmode
+	element_flags = ELEMENT_BESPOKE
+	argument_hash_start_idx = 1
+
+	/// The type of area that will trigger godmode.
+	var/area_type
+
+	/// Whether or not to allow subtypes of the area type to trigger godmode.
+	var/allow_area_subtypes
+
+	/// The message to send to the mob when they gain godmode.
+	var/gain_message
+
+	/// The message to send to the mob when they lose godmode.
+	var/lose_message
+
+	/// Lazy. The cache of mobs that are currently in godmode; prevents sending the gain message multiple times.
+	var/list/godmode_cache
+
+/datum/element/area_based_godmode/Attach(
+	datum/target,
+	area_type,
+	allow_area_subtypes = FALSE,
+	gain_message = span_big(span_green("You are now invulnerable.")),
+	lose_message = span_big(span_red("You are no longer invulnerable.")),
+)
+	. = ..()
+
+	if(!ismob(target))
+		return ELEMENT_INCOMPATIBLE
+
+	var/mob/mob_target = target
+	if(initial(mob_target.status_flags) & GODMODE)
+		return ELEMENT_INCOMPATIBLE
+
+	src.area_type = area_type
+	src.allow_area_subtypes = allow_area_subtypes
+	src.gain_message = gain_message
+	src.lose_message = lose_message
+
+	mob_target.become_area_sensitive(type)
```

Comment:
should use REF(src) here, since you could have more then one source of area godmode at once right?